repoaudit/repoaudit/yum.py

```diff
@@ -14,23 +20,127 @@
+                file_checksum_info = repomd.find(f"repo:data[@type='{data_type}']/repo:checksum",
+                                                 namespaces=NS)
 
-        packages = primary.findall("common:package", namespaces=NS)
+                if file_location_info is None or file_checksum_info is None:
+                    errors.add(
+                        url, RepoErrors.YUM_DIST,
+                        f"{repomd_url} file malformed, "
+                        f"no location or checksum found for {data_type}"
+                    )
+                    raise ParseError
+
+                file_loc = file_location_info.get("href")
+                checksum_type = file_checksum_info.get("type")
+
+                if file_loc is None or checksum_type is None:
+                    errors.add(
+                        url, RepoErrors.YUM_DIST,
+                        f"{repomd_url} file malformed"
+                        f"no href or type for {data_type}"
+                    )
+                    raise ParseError
+
+                file_url = urljoin(url, file_loc)
+
+                if checksum_type == "sha":
+                    checksum_type = "sha1"
+                multihash = MultiHash([checksum_type])
+
+                ref_checksum = file_checksum_info.text
+
+                try:
+                    response = get_url(f"{file_url}", verify=verify, stream=True)
+                except HTTPError as e:
+                    errors.add(
+                        url, RepoErrors.YUM_DIST,
+                        f"Could not access file at {e.response.url}: {e}"
+                    )
+                    success = False
+                    continue
+
+                for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
+                    multihash.update(chunk)
+
+                if multihash.hexdigest(checksum_type) != ref_checksum:
+                    errors.add(
+                        url, RepoErrors.YUM_DIST,
+                        f"Metadata {checksum_type} checksum mismatch for '{file_url}'. Expected "
+                        f"'{ref_checksum}' but received '{multihash.hexdigest(checksum_type)}'."
+                    )
+                    success = False
+    except ParseError:
+        success = False
+
+    if success:
+        click.echo("Metadata check successful")
+    else:
+        click.echo("Metadata check failed")
+
+
+def _check_yum_signature(url: str, gpg: Optional[gnupg.GPG],
+                         errors: RepoErrors, verify: Optional[str] = None) -> None:
+    """Verify signature using provided public keys in gpg parameter."""
+    if gpg is None:
+        return
+
+    repomd_url = urljoin(url, "/repodata/repomd.xml")
+    repomdsig_url = urljoin(url, "/repodata/repomd.xml.asc")
+
+    success = check_signature(url, RepoErrors.YUM_DIST, repomd_url, gpg, errors,
+                              signature_url=repomdsig_url, verify=verify)
+
+    if "suse" in url or "sles" in url:
```

Comment:
So a part of this check is making sure that the key actually exists. Many suse repo's were missing this key when they were first checked but they have been fixed since. I think another method is needed for determining if the repo is a suse repo. Any suggestions?

Reply:
> So a part of this check is making sure that the key actually exists. Many suse repo's were missing this key when they were first checked but they have been fixed since.

I see, that makes sense.

> I think another method is needed for determining if the repo is a suse repo. Any suggestions?

I think this is fine for now. Let's a file an issue and continue discussion there.